grafonnet/gauge_panel.libsonnet

```diff
@@ -0,0 +1,193 @@
+   * @param showThresholdMarkers Render the thresholds as an outer bar.
+   * @param unit Panel unit field option.
+   * @param min Leave empty to calculate based on all values.
+   * @param max Leave empty to calculate based on all values.
+   * @param decimals Number of decimal places to show.
+   * @param displayName Change the field or series name.
+   * @param noValue What to show when there is no value.
+   * @param thresholdsMode 'absolute' or 'percentage'.
+   * @param repeat Name of variable that should be used to repeat this panel.
+   * @param repeatDirection 'h' for horizontal or 'v' for vertical.
+   * @param repeatMaxPerRow Maximum panels per row in repeat mode.
+   * @param pluginVersion Plugin version the panel should be modeled for. This has been tested with the default, '7', and '6.7'.
+   *
+   * @method addTarget(target) Adds a target object.
+   * @method addTargets(targets) Adds an array of targets.
+   * @method addLink(link) Adds a link. Aregument format: `{ title: 'Link Title', url: 'https://...', targetBlank: true }`.
+   * @method addLinks(links) Adds an array of links.
+   * @method addThreshold(step) Adds a threshold step. Aregument format: `{ color: 'green', value: 0 }`.
+   * @method addThresholds(steps) Adds an array of threshold steps.
+   * @method addMapping(mapping) Adds a value mapping.
+   * @method addMappings(mappings) Adds an array of value mappings.
+   * @method addDataLink(link) Adds a data link.
+   * @method addDataLinks(links) Adds an array of data links.
+   */
+  new(
+    title,
+    description=null,
+    transparent=false,
+    datasource=null,
+    allValues=false,
+    valueLimit=null,
+    reducerFunction='mean',
+    fields='',
+    showThresholdLabels=false,
+    showThresholdMarkers=true,
+    unit='percent',
+    min=0,
+    max=100,
+    decimals=null,
+    displayName=null,
+    noValue=null,
+    thresholdsMode='absolute',
+    repeat=null,
+    repeatDirection='h',
+    repeatMaxPerRow=null,
+    pluginVersion='7',
+  ):: {
+
+    type: 'gauge',
+    title: title,
+    [if description != null then 'description']: description,
+    transparent: transparent,
+    datasource: datasource,
+    targets: [],
+    links: [],
+    [if repeat != null then 'repeat']: repeat,
+    [if repeat != null then 'repeatDirection']: repeatDirection,
+    [if repeat != null then 'repeatMaxPerRow']: repeatMaxPerRow,
+
+    // targets
+    _nextTarget:: 0,
+    addTarget(target):: self {
+      local nextTarget = super._nextTarget,
+      _nextTarget: nextTarget + 1,
+      targets+: [target { refId: std.char(std.codepoint('A') + nextTarget) }],
+    },
+    addTargets(targets):: std.foldl(function(p, t) p.addTarget(t), targets, self),
+
+    // links
+    addLink(link):: self {
+      links+: [link],
+    },
+    addLinks(links):: std.foldl(function(p, l) p.addLink(l), links, self),
+
+    pluginVersion: pluginVersion,
+  } + (
+
+    if pluginVersion >= '7' then {
```

Comment:
What do you think about `{ } + (import 'gauge_panel_mixin_%s.libsonnet' % pluginVersion)`? 

If/else is not very jsonnet idiomatic, not sure how and if we should resolve that here or in a later version of Grafonnet. @malcolmholmes WDYT?

Reply:
`pluginVersion` isn't quite so deterministic. It corresponds to the minimum version of Grafana being targeted. Folks could put `7.0.5`, `6.7.4`, `6.6`, `5`, etc. which wouldn't always match a file. But we could do something similar.

I had considered this after first implementing support for 7.0 and then adding backwards compatibility for 6.7. I chose this strategy because it keeps everything contained within a single file. Having an object span multiple files would start a new convention for version support. And this is something I didn't want to do just yet because I think it needs more thought and it's going to be relevant for all objects.

The approach I had in mind uses top-level directories for the specific versions supported:

```
grafonnet/
7.0/
6.7/
6.6/
...
```

This would allow users to specify a version when they import Grafonnet as opposed to every time they instantiate an object.

Reply:
... or in a later version of Grafonnet. :soon: